Driver for LX Navigation Eos57

jirimaier opened the pull request:
<!--

Thank you for your interest in contributing to XCSoar! Please read the
following information to make it easier for us to review your changes.

We appreciate if you make sure to:

  - Document the changes (in the NEWS.txt file, and the manual when relevant)
  - Enable maintainer edits[1] (in case we need to help with the PR)
  - Check your commits and their messages (so they're all tidy and coherent)

[1] https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->


Brief summary of the changes
----------------------------
Driver for LX Navigation Eos57 variometer.

Features
----------------------------
- Receiving NMEA data, correctly interpreting altitude
- Task declaration, including observation zones shape
- Flight log download
- MacCready and bugs and ballast synchronization

Compatibility
----------------------------
- Tested on Eos57
- Might work well with Era and Eos80 too. That is not tested, and it is possible that some features will not work properly. Anyway, it will most likely work better than the LXNAV driver. 

Known issues
----------------------------
- Declaration does not set the Landing entry in the task settings. This behavior is the same as with the LXNAV driver and is probably caused by a bug in Eos's firmware. It does not pose any problem for usage, as the Landing point doesn't have any practical usage.
- Ballast synchronization may not be precise. The value provided by vario is relative to polar reference mass. The reference mass of vario's polar (can be different from XCSoar's polar) cannot be directly obtained and is estimated from polar coefficients. 

<!--
Please note that the commit messages is where detailed descriptions of the
changes should be made - this section is just for a brief summary/overview.
-->


Related issues and discussions
------------------------------
New generation LX Navigation Device #389
Wrong altitude from Lx Eos variometer #969
ERA80-Xcsoar #415
<!--
Please link any relevant issues or forum posts here, for reference.

If this PR resolves an existing issue, please write "Closes #1234" so that
the issue is closed automatically when this PR is merged.
-->

MaxKellermann commented:
This PRs copies lots of existing code. That is not acceptable and will not get merged.

jirimaier commented:
Thank you for the feedback.

I modified the code to call the LXNAV driver's methods for parsing LXWP1 and LXWP3 sentences instead of copying the code (I think this was the main issue with copying existing code).

I also modified the tests to use an LXWP1 sentence which is given as an example in the LX Navigation Communication Protocol documentation (instead of copying the one from the LXNAV driver test).

I think that now there is no copied code. Some parts may be similar because the protocol is similar, but not exactly the same. I don't think that I could reuse other code from the LXNAV driver without making modifications to that driver. 

The Codacy Analysis shows issues regarding unused variables in src/Device/RecordedFlight.hpp, I believe these are false positives. These variables are used in DownloadFlight and GetFlightInfo methods of LXEosDevice.

lordfolken commented:
This ignores the vario's qnh setting. 
We can hope that somehow xcsoars autoqnh and the varios undocumented autoqnh somehow arrive at the same conclusion.

lordfolken commented:
Also please test the driver post launch detection. I.e. with pressure on the tas. The vario behaves different post launch.

jirimaier commented:
> This ignores the vario's qnh setting. We can hope that somehow xcsoars autoqnh and the varios undocumented autoqnh somehow arrive at the same conclusion.

The QNH of XCSoar is set based on the "alt_offset" from LXWP3 sentence.

I am not sure what you mean by "varios autoqnh". When the vario is turned on, the user sets the current altitude (airport elevation), and the QNH of the vario is set accordingly. The altitude offset in LXWP3 can then be converted to the QNH setting so that the QNH in XCSoar gets set to the same value as on the vario.

lordfolken commented:
> > This ignores the vario's qnh setting. We can hope that somehow xcsoars autoqnh and the varios undocumented autoqnh somehow arrive at the same conclusion.
> 
> The QNH of XCSoar is set based on the "alt_offset" from LXWP3 sentence.
> 
> I am not sure what you mean by "varios autoqnh". When the vario is turned on, the user sets the current altitude (airport elevation), and the QNH of the vario is set accordingly. The altitude offset in LXWP3 can then be converted to the QNH setting so that the QNH in XCSoar gets set to the same value as on the vario.

The alt_offset will remain the same value, after launch detection on the nmea side, no matter what you set the qnh on the varios ui to.

jirimaier commented:
I will test the in-flight behavior soon (probably tomorrow).

jirimaier commented:
I never changed the QNH on the vario during flight.
If I understand the behavior of the vario correctly:
1) After turning on the vario, pilot sets the elevation.
2) Vario calculates QNH based on the elevation
3) Pilot can change the QNH value, but that doesn't change the altitude (which is intentional, because you just want to tweak the QNH while keeping the elevation that you set)
4) After confirming this settings, the elevation cannot be changed, but it is possible to change the QNH in settings.
5) Changing QNH before takeoff has no effect on altitude (same as in initial setup)
6) Changing QNH in-fligh will change the true altitude 

And you say that the alt_offset doesn't change when the QNH is changed in flight (This is a bug in the vario, right?).

So this means that if the pilot changes QNH on the vario during flight, they also have to change QNH in the XCSoar. This is not so big problem I guess. I just need to make sure that the new (manually set) QNH in XCSoar doesn't get overwritten to the old value by the next LXWP3 sentence (that has outdated alt_offset).

lordfolken commented:
That summarizes it pretty well. 
Yes, its a bug at least in the era i have.
Basically stop accepting the alt_offset post takeoff.

jirimaier commented:
I tested it in flight today. 

When changing QNH on the vario, the QNH in XCSoar also got changed. 

So it seems that the bug is not present in Eos. 

The baro altitude and FL shown by XCSoar always matched the values shown by vario, so the testing is successful. 

To make it compatible with the bug in Era, I can modify the LXWP3 parsing so that the QNH will be set only if the alt_offset is different from the one received in previous LXWP3 (or if it is the first sentence since connecting). 
That way, change in alt_offset will cause change of QNH, but constant value will be ignored. It wouldn't affect the behavior in Eos, and also would fix the problem with overwriting new QNH se by user in case of Era. 

Anyway, I noticed that if I set the QNH in XCSoar manually, all changes of QNH in the vario are ignored.
Is this intended behavior of XCSoar (that it ignores attempts to change QNH from the driver once the QNH was changed by user)?

jirimaier commented:
It is a bit strange that in the initial setup of the vario, the elevation and QNH can be set independently. In XCSoar, the QNH and the difference between true and standard altitude are totaly dependent on each other. 

The alt_offset in LXWP3 only depends on the initial elevation and on in-flight changes of QNH, it doesn't depend on the initial QNH.

If the initial QNH on vario is changed to different value than the one calculated from elevation (for whatever reason), the QNH in XCSoar (based on alt_offset) will be different from the QNH in vario. But this is fine, because the QNH in XCSoar will lead to correct conversion of true altitude to standard altitude,as it is based on the alt_offset.

jirimaier commented:
I noticed one more issue:

If the Eos is connected together with another device (Flarm, for example), the altitude provided by the other dece is prioritized regardless of the order in the device list.

In other words: the info.ProvideBaroAltitudeTrue(altitude_true) is "weaker" than info.ProvidePressureAltitude(value).

Can this be considered a bug in XCSoar, or is it a feature?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.hpp:
```diff
@@ -16,6 +16,85 @@
+  uint8_t prg[12];
+
+  // TP Longitude in degrees multiplied by 60000.0f
+  int32_t lon[12];
```
What's the byte order?

jirimaier replied:
big-endian, I added that to the comment

MaxKellermann replied:
Then why not use a safe class like PackedBE32 here? That way, it's impossible to misuse it, and you don't even need a comment.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.hpp:
```diff
@@ -16,6 +16,85 @@
 
 using std::string_view_literals::operator""sv;
 
+struct EosDeclarationStruct
```
This should have a `static_assert` checking the size of the struct at compile time to verify its correctness

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.hpp:
```diff
@@ -16,6 +16,85 @@
+
+  uint8_t bAutoNext = 1; // Is this auto next TP or AAT TP
+  uint8_t bIsLine = 0;   // Is this line flag
+  float fA1;             // Angle A1 in radians
```
Byte order? What bit format is being used? Is this bit format guaranteed on all CPU architectures?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.cpp:
```diff
@@ -137,6 +137,68 @@ LXEosDevice::SendNewSettings(OperationEnvironment& env)
+}
+
+uint8_t
+LXEosDevice::CalculateCRC(std::byte* msg, const int len, const uint8_t initial)
+{
+  uint8_t result{ initial };
+
+  for (int byte = 0; byte < len; byte++) {
+    uint8_t d = static_cast<uint8_t>(msg[byte]);
+    for (int count = 8; --count >= 0; d <<= 1) {
+      uint8_t tmp = result ^ d;
+      result <<= 1;
+      if ((tmp & 0x80) != 0)
+        result ^= 0x69;
+    }
+  }
+
+  return result;
+}
```
Duplicate code, already exists in the LXNAV driver

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.hpp:
```diff
@@ -95,6 +97,34 @@ struct EosClassStruct
+  const uint8_t cmd = 0xF1;
+
+  // Flight ID is the index (1 = newest), not the ID from FlightInfo
+  uint16_t flight_id;
```
Byte order?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,270 @@
+}
+
+bool
+LXEosDevice::GetFlightLogBlock(AllocatedArray<std::byte>& block,
```
This function should return the `AllocatedArray` instead of passing it by reference; throw an exception on error.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,270 @@
+  flight.internal.lx_eos.flight_id = index;
+
+  uint32_t julianDate, time_takeoff, time_landing, file_size;
+  std::memcpy(&julianDate, &response[13], 4);
+  std::memcpy(&time_takeoff, &response[17], 4);
+  std::memcpy(&time_landing, &response[21], 4);
+  std::memcpy(&file_size, &response[89], 4);
```
Byte order? Why are you using a byte array with hard-coded offsets here, instead of declaring a (documented) struct, as you did in other parts of this PR?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,270 @@
+                  env,
+                  std::chrono::milliseconds(5000));
+  } catch (...) {
+    return false;
```
What's the point of catching this exception? You discard all information about the failure by converting it to`false` here.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,270 @@
+  uint8_t crc = CalculateCRC(responseHeader, sizeof(responseHeader), 0xFF);
+  crc = CalculateCRC(block.data(), block.size(), crc);
+  if (CalculateCRC(crc_byte, 1, crc) != 0)
+    return false;
```
Throw exception instead of returning false

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,270 @@
+}
+
+BrokenDate
+LXEosDevice::julianToDate(uint32_t julian_date)
```
I think it could be useful to move this to class BrokenDate, e.g. `BrokenDate::FromJulianDay(unsigned)`.

MaxKellermann commented on src/Device/RecordedFlight.hpp:
```diff
@@ -49,6 +49,14 @@ struct RecordedFlightInfo : FlightInfo {
     unsigned flytec;
 
+    struct {
+      /**
+       * Flight id and igc file size, used by the LX Eos driver.
+       */
```
A doxygen comment documents only the following field, not both. Split this comment.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDeclare.cpp:
```diff
@@ -0,0 +1,218 @@
+   * bytes are used for the angles */
+
+  // Bytes of float values for 45, 90 and 180 degree angles in radians
+  uint8_t angle180[] = { 0xdb, 0x0f, 0x49, 0x40 };
+  uint8_t angle90[] = { 0xdb, 0x0f, 0xc9, 0x3f };
+  uint8_t angle45[] = { 0xdb, 0x0f, 0x49, 0x3f };
```
This is extremely obscure and for no good reason, it allocates 12 bytes on the stack at runtime on every call and initializes them. If you really need constants like that, make them `static constexpr`, but I doubt this is the right way to do it.

jirimaier replied:
Can I assume that all platforms use IEEE 754 float? How is float affected by big/little endian?

MaxKellermann replied:
> Can I assume that all platforms use IEEE 754 float?

You need to check https://en.cppreference.com/w/cpp/types/numeric_limits/is_iec559 - I agree to make this a `static_assert`, until we see that a currently supported target doesn't support it.

> How is float affected by big/little endian?

I don't know, I never cared. But if you write code that transmits floats in a binary stream, it's up to you to figure out what byte order (and bit format) is expected/emitted by the device you're talking to, and you must figure out the byte order (and bit format) of all CPUs supported by XCSoar.

As I wrote in another comment, this shouldn't be sprinkled everywhere, but implemented in one central type, like I did with classes like `PackedLE32`.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDeclare.cpp:
```diff
@@ -0,0 +1,218 @@
+  /* We cannot use predefined bytes for elevation (it can be any value)
+   * Check if the float format on this platform matches the expected one
+   * by comparing with 4 bytes oof known-good value */
+  float test_float = 123.45e33;
+  uint8_t expected_bytes[] = { 0x70, 0x34, 0xbe, 0x79 };
+  if (memcmp(&test_float, &expected_bytes, 4) == 0) {
+    // Yes
+    data.fElevation = declaration.GetWaypoint(tp_nr - 1).GetElevationOrZero();
+  } else {
+    // No. Maybe the bytes are swapped?
+    uint32_t test_float_bytes = 0;
+    memcpy(&test_float_bytes, &test_float, 4);
+    test_float_bytes = ByteSwap32(test_float_bytes);
+    if (memcmp(&test_float_bytes, &expected_bytes, 4) == 0) {
+      // Use swapped bytes
+      float elev = declaration.GetWaypoint(tp_nr - 1).GetElevationOrZero();
+      uint32_t elev_bytes = 0;
+      memcpy(&elev_bytes, &elev, 4);
+      elev_bytes = ByteSwap32(elev_bytes);
+      memcpy(&data.fElevation, &elev_bytes, 4);
+    } else {
+      // Set zero if the format is something exotic
+      memset(&data.fElevation, 0, 4);
+    }
+  }
```
All of this is ... awful. This adds runtime overhead by forcing the compiler to generate memcmp() calls, even on CPUs where nothing needs to be changed. All of these checks should be at compile-time and generate only runtime code if runtime code is really necessary. And it should be encapsulated in a generic class that can be reused everywhere. Solve this once, properly, and then reuse this class everywhere you need a float in a specific bit format with a specific byte order.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.cpp:
```diff
@@ -133,6 +133,50 @@ LXEosDevice::SendNewSettings(OperationEnvironment& env)
+void
+LXEosDevice::CopyStringSpacePadded(char dest[],
+                                   const TCHAR src[],
+                                   const uint8_t len)
```
A buffer length declared as `uint8_t`? Huh? `std::size_t`!

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.hpp:
```diff
@@ -173,14 +250,14 @@ class LXEosDevice : public AbstractDevice
+   * @param polar2
+   * @return true: coefficients are the same
+   * @return false: coefficients are different or at least one polar is invalid
+   */
```
Why change indentation here?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.hpp:
```diff
@@ -247,8 +324,7 @@ class LXEosDevice : public AbstractDevice
    * @param env
-   * @return true
-   * @return false
+   * @return true if successful
```
?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.hpp:
```diff
@@ -259,11 +335,47 @@ class LXEosDevice : public AbstractDevice
    * @return true if ACK was received,
    * @return false if time out or other exception occurred
-   * @note timeout is 1 second
+   * @note timeout is 3 seconds
```
?

jirimaier replied:
It returns false if ACK is not received within 3 seconds

MaxKellermann replied:
But why is the documentation changed in this commit? Did you change the timeout? Why?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.hpp:
```diff
@@ -259,11 +335,47 @@ class LXEosDevice : public AbstractDevice
+   * @return true if successful
+   * @note(flight_count not changed on unsuccess)
+   */
+  bool GetNumberOfFlights(uint8_t& flight_count, OperationEnvironment& env);
```
Return the number of flights instead of passing a reference variable for the result.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+#include "Device/Error.hpp"
```
License header is missing

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+      RecordedFlightInfo flight;
+
+      // Try up to 5 times before giving up
+      for (uint8_t attempt = 0; attempt < 5; attempt++) {
```
Are you aware that declaring this `uint8_t` is not faster, but actually adds overhead? It's a fallacy.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+      try {
+        block = GetFlightLogBlock(flight_id, block_id, env);
+      } catch (...) {
+        success = false;
```
This discards details about the error condition. Reporting error details to the user (or the log file) is very useful, don't throw that away!

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+{
+  // Request
+  EosGetNumOfFlights data;
+  auto data_array = reinterpret_cast<std::byte*>(&data);
+  std::span<std::byte> message(data_array, sizeof(data));
```
Complicated and fragile way of casting to a span.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+      env,
+      std::chrono::milliseconds(5000));
+  } catch (...) {
+    return false;
```
Again: what's the point of this exception handler All it does is discard details about the error condition, but why?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+      env,
+      std::chrono::milliseconds(5000));
+  } catch (...) {
+    return false;
```
Again: what's the point of this exception handler All it does is discard details about the error condition, but why?

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+  if (Calculate8bitCRC(
+        reinterpret_cast<uint8_t*>(&info), sizeof(info), 0xFF, 0x69) !=
+      0x00) // Check CRC
+    return false;
```
Throw exception instead of returning false with no details.

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+                               uint16_t block_id,
+                               OperationEnvironment& env)
+{
+  AllocatedArray<std::byte> block;
```
Declare this variable where it is first used; declaring it here bloats the code and adds runtime overhead!

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+  data.block_id = ToLE16(block_id);
+  data.crc = Calculate8bitCRC(
+    reinterpret_cast<uint8_t*>(&data), sizeof(data) - 1, 0xFF, 0x69);
+  std::span<std::byte> message(reinterpret_cast<std::byte*>(&data),
+                               sizeof(data));
```
Clumsy span cast

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDownload.cpp:
```diff
@@ -0,0 +1,245 @@
+    reinterpret_cast<uint8_t*>(&header), sizeof(header), 0xFF, 0x69);
+  crc = Calculate8bitCRC(block.data(), block.size(), crc, 0x69);
+  if (Calculate8bitCRC(crc_byte, 1, crc, 0x69) != 0)
+    throw("Bad CRC");
```
Don't throw a C string; also the parentheses are unnecessary. `throw` is not a function.

MaxKellermann commented on src/Device/RecordedFlight.hpp:
```diff
@@ -49,8 +49,16 @@ struct RecordedFlightInfo : FlightInfo {
+    /**
+     * Flight id and igc file size, used by the LX Eos driver.
+     */
+    struct {
+      uint16_t flight_id;
+      uint32_t file_size;
+    } lx_eos;
+
   } internal;
 };
 
-class RecordedFlightList : public StaticArray<RecordedFlightInfo, 128u> {
+class RecordedFlightList : public StaticArray<RecordedFlightInfo, 128u>{
```
??

jirimaier commented on src/Device/Driver/LX_EOS/LXEosDeclare.cpp:
```diff
@@ -0,0 +1,218 @@
+  uint8_t angle45[] = { 0xdb, 0x0f, 0x49, 0x3f };
+
+  // Init A21 and A2 to zero
+  memset(&data.fA21, 0, 4);
+  memset(&data.fA2, 0, 4);
```
4 is the size of float in the structure that is transmitted to the vario (if it wasn't, the static_assert of structure size would fail). 
In the previous review, you pointed out that the number format may not be guaranteed on all architectures, so I wanted to explicitly set all bytes to zero rather than expecting that float zero is represented like that on any platform.

MaxKellermann replied:
This is all true, but it has nothing to do with my complaint. In this line of code, you don't see what `fA21` is, and you don't know how big it is. This obscures the code. You could have used `sizeof(data.fa21)` or just do `data.fA21 = {};`, but you didn't - instead you hard-coded `4` which makes your code rather obscure.

jirimaier commented on src/Device/Driver/LX_EOS/LXEosDevice.cpp:
```diff
@@ -133,6 +133,50 @@ LXEosDevice::SendNewSettings(OperationEnvironment& env)
+      std::byte{ 0x06 },
+      env,
+      std::chrono::milliseconds(3000));
+  } catch (...) {
+    return false;
```
When calling the method, I am interested in whether the transmission was successful (ACK was received afterward) or not. Why should I keep the details?

MaxKellermann replied:
To differentiate between "there was no reply at all" or "there was a malformed reply, what the hell is this device saying?" or "there was a nack".

MaxKellermann commented on src/Device/Driver/LX_EOS/LXEosDevice.cpp:
```diff
@@ -133,6 +133,50 @@ LXEosDevice::SendNewSettings(OperationEnvironment& env)
+  try {
+    port.WaitForByte(
+      // ACK is 0x06 (NACK is 0x15)
+      std::byte{ 0x06 },
```
There should be a central protocol definitions giving these magic values a good name. That way, you can avoid adding those comments everywhere. Look at  the LXNAV driver, it already has such a definition! And the Volkslogger driver has another one. Bad duplicate code, of course, but still better than your code.

MaxKellermann commented on src/util/ByteOrder.hxx:
```diff
@@ -263,7 +264,7 @@ public:
 		return x;
 	}
-};
+} __attribute__((packed));
```
This is a proprietary compiler extension that should not be used unconditionally, but why use it here??

MaxKellermann replied:
Also, adding it to this existing structure is unrelated to this commit and is not even explained in the commit message.
Please don't hide unrelated changes. I'll find crap like that!

jirimaier replied:
When I used the PackedLE32 (or others) in my structures (which are using this attribute to ensure no padding), the compiler gave error that the attribute (of the struct) is ignored because it contains "unpacked non-POD". Adding the attribute to the classes fixed that.

MaxKellermann replied:
So you have no idea what you're doing, and you did some random thing and this made the compiler shut up, so you figured this must be the solution to your problem?

jirimaier replied:
The __attribute__((packed)) is used in an example of data structure in the protocol documentation from LX Navigation. So it is not entirely random. But you are right that I have no deep understanding of what this does, just that it somehow ensures that there is no padding between the variables.

jirimaier replied:
I previously used this for my structures in the LXEosDevic, which was already reviewed by you I think.

jirimaier replied:
Should I rather use gcc_packed (as in the LXNAV driver)? What is the difference?

jirimaier replied:
The LXNAV drive is using the unit and int types in the structures, not the PackedBE32 and others. I think they are actually used nowhere in the code.

MaxKellermann replied:
> The __attribute__((packed)) is used in an example of data structure in the protocol documentation from LX Navigation.

And LX Navigation's documentation follows which standard?

> So it is not entirely random.

You're missing the point, and it sounds like you don't know what I'm talking about. You havn't even tried to answer my question. I asked why you use it. You have not explained.

Saying it's in the LX Navigation documentation in example code doesn't answer the question, unless you want to tell me that you just copy somebody else's code without understanding it. Code that nobody can explain will never be merged into XCSoar.

> Should I rather use gcc_packed (as in the LXNAV driver)?

Why would you want to do that? This simple question, you're still sidestepping it wherever you can, instead of answering it.

jirimaier replied:
My understanding is: When I have a structure that contains variables with sizes different than 4 bytes (on 32-bit platform), the compiler may decide to pad the less-than-4-byte variable with empty bytes so that the next variable begins at an address that is multiple of 4, because the CPU can work with that efficiently. This is unwanted for my data structures, because it must exactly match the communication protocol bytes. I already wrote 

> it somehow ensures that there is no padding between the variables

I know about two ways to achieve this: gcc_packed and the attribute((packed)) 
I don't know which of them is preferred. 

And it seems that the attribute cannot work on a struct that contains "unpacked non-POD", so I added the attribute to the PackedBE32 and others too. It makes no practical difference for them (if they had any padding, the static_assert of their size would fail), but it 

> made the compiler shut up

jirimaier replied:
BTW the reason I chose atribute((packed)) rather than gcc_packed is that when using gcc_packed, my IDE was complaining about the asserts checking the size of the structs. Of course, this is just some wrong setting of the IDE, which is not a good reason for deciding which to use. But I wasn't aware of any reason to prefer one over the other anyway.

MaxKellermann replied:
Your understanding is not correct.

jirimaier replied:
You can help me better understand it by telling me if the following claims are true or not:

The compiler can place unused bytes between variables to ensure that variables are properly aligned in the memory.

Aligned means that they can be accessed efficiently. Misaligned causes lower performance or even exceptions.

On a 32-bit platform, properly aligned variables will start at a memory address that is multiple of 4

The data structure used for the communication protocol must not contain any padding.

This can be achieved by telling the compiler not to do any alignment in the structure.

Which is the purpose of the "packed" attribute.

Structure with this attribute cannot have "unpacked non-POD" members.

These are classes or structures that do not have the packed attribute.

So I need to add the attribute to the Packed32BE (and other types I use)

MaxKellermann replied:
No, you can't just put the burden on me and throw this wall of statements at me, making it my responsibility to verify them. I don't have time for that.
And no again, PackedBE32 doesn't need the "padded" attribute.
I'm tired of this PR, I'm going to unsubscribe. When you're completely done, have fixed all the problems and it's no longer in "Draft" status, you may ask for another review, but until then, I decide my time is too precious for you to waste it.

jirimaier replied:
Honestly, I was not expecting you to review it while it is still in "draft" status...

jirimaier replied:
I still don't understand what is going on

The Packed32BE indeed is 4 byte even if the "packed" attribute is not used. 

However, when I try to use the Packed32BE in the struct of data that are transmitted to vario (this struct obviously must be packed), the compiler will say that the packed attribute is ignored because the structure contains "unpacked non-POD". 

Adding the packed attribute to Packed32BE solves that, but you say that it is not correct (or at least not needed) to do so. So what would be a good solution? The Packed32BE is not used anywhere else in the code; I thought this was the intended use-case of Packed32BE, was it? Was it ever used (and working)?

jirimaier replied:
I still think that the attribute needs to be there (to be able to use the Packed32BE inside another struct that is packed)

jirimaier replied:
So if I try to answer my previous questions:

> The compiler can place unused bytes between variables to ensure that variables are properly aligned in the memory.

True (or at least close-enough to true)

> Aligned means that they can be accessed efficiently. Misaligned causes lower performance or even exceptions.

True (or at least close-enough to true)

> On a 32-bit platform, properly aligned variables will start at a memory address that is multiple of 4

False, it is a multiple of the size of the data type, so for example `uint8_t a,b,c,d;` will have no padding.

> The data structure used for the communication protocol must not contain any padding.

True

> This can be achieved by telling the compiler not to do any alignment in the structure.
> Which is the purpose of the "packed" attribute.

True (or at least close-enough to true)

> Structure with this attribute cannot have "unpacked non-POD" members.

That's what the compiler is complaining about

> These are classes or structures that do not have the packed attribute.

Maybe false, but seems like the compiler will be satisfied only if I add the attribute

> So I need to add the attribute to the Packed32BE (and other types I use)

That's what I think. If it is not true, I would be glad for any suggestions on how to solve it in other way.

MaxKellermann commented on src/util/ByteOrder.hxx:
```diff
@@ -516,7 +517,55 @@ public:
+
+  PackedFloatLE(float f)
+  {
+    value.floating_point = f;
```
This will crash the process with SIGBUS on certain CPUs.

MaxKellermann commented on src/util/ByteOrder.hxx:
```diff
@@ -516,7 +517,55 @@ public:
+  {
+    value.floating_point = f;
+    if (IsBigEndian()) {
+      value.unsigned_int = ByteSwap32(value.unsigned_int);
```
This will also crash.

MaxKellermann commented on src/util/ByteOrder.hxx:
```diff
@@ -516,7 +517,55 @@ public:
+{
+  union Value
+  {
+    uint32_t unsigned_int = 0;
```
What's the point of this initializer?

MaxKellermann commented on src/util/ByteOrder.hxx:
```diff
@@ -516,7 +517,55 @@ public:
+public:
+  PackedFloatLE() = default;
+
+  PackedFloatLE(float f)
```
Your implementation is not `constexpr`, that's bad.

MaxKellermann commented on src/util/ByteOrder.hxx:
```diff
@@ -516,7 +517,55 @@ public:
+  }
+} __attribute__((packed));
+
+static_assert(std::numeric_limits<float>::is_iec559);
```
Why this? I know I suggested doing that, but the next guy won't understand it. And you didn't even document this requirement in the class's documentation. Oh, there's no API documentation at all!

MaxKellermann commented on src/util/ByteOrder.hxx:
```diff
@@ -516,7 +517,55 @@ public:
+} __attribute__((packed));
+
+static_assert(std::numeric_limits<float>::is_iec559);
+static_assert(sizeof(PackedFloatLE) == sizeof(uint32_t), "Wrong size");
```
This is an obscure requirement. Why should the size of this class be the same as `uint32_t`? It's not a 32-bit integer. It's a 32-bit float. Sure, in the end, both are 4 bytes, but you're not testing for it to be 32 bit or 4 bytes, you're testing this to be uint32_t, but why this choice? This is confusing.

jirimaier replied:
I know for sure that the result (to be sent to the vario) should be 4 bytes. What if the float is not 4 bytes on some platform? Maybe it can't be, I don't know. But it was the reason why I did it like this.

jirimaier replied:
However, when the std::numeric_limits<float>::is_iec559 is true, it probably guarantees that sizeof(float)==4

MaxKellermann replied:
Sigh. I already said that you're "**not testing for it to be 32 bit or 4 bytes**". Now you say you do it because you want to test whether it's 4 bytes. This sounds like you didn't read my comment.

MaxKellermann replied:
> However, when the std::numeric_limits::is_iec559 is true, it probably guarantees that sizeof(float)==4

"Probably" isn't good enough, even less so if you're wrong.

MaxKellermann commented on src/util/ByteOrder.hxx:
```diff
@@ -516,7 +517,61 @@ public:
+  constexpr PackedFloatLE(float f)
+  {
+    uint32_t value_as_uint;
+    value_as_uint = *reinterpret_cast<uint32_t*>(&f);
```
This is the worst possible way to express this because the compiler is forced to allocate stack space for the parameter (cannot be an a register) and it violates C++'s aliasing rules, so it's UB.

lordfolken commented on src/Device/Driver/LX_EOS/LXEosDevice.cpp:
```diff
@@ -0,0 +1,199 @@
+LXEosDevice::HasReliableAltOffset(DeviceInfo device)
+{
+  if (device.product == "LX Eos"sv) {
+    if (device.software_version == "1.7"sv)
```
Only Version 1.7 or equal or bigger then?

lordfolken replied:
I would even consider refusing service on a non supported firmware, instead of working around bugs on the device.

lordfolken commented on src/Device/Driver/LX_EOS/LXEosDevice.cpp:
```diff
@@ -0,0 +1,199 @@
+}
+
+bool
+LXEosDevice::PutBallast([[maybe_unused]] double fraction,
```
Beware, I think xcsoar is currently sending the wrong ratio to the external devices. Its falsley using dry mass instead of polar reference mass while talking to external devices. See my overload2 PR.

lordfolken replied:
Internally its using polar reference mass correctly.

lordfolken commented on src/Device/Driver/LX_EOS/LXEosRegister.cpp:
```diff
@@ -13,6 +13,7 @@ LXEosCreateOnPort([[maybe_unused]] const DeviceConfig& config,
+  DeviceRegister::DECLARE | DeviceRegister::RECEIVE_SETTINGS |
+    DeviceRegister::SEND_SETTINGS,
+  LXEosCreateOnPort
 };
```
Missing new line

lordfolken commented on src/Device/Driver/LX_EOS/LXEosDeclare.cpp:
```diff
@@ -0,0 +1,174 @@
+  CopyStringSpacePadded(data.name, _T(""), sizeof(data.name));
+
+  // CRC is calculated only from data bytes (excluding the SYN and CMD)
+  // Using poly 0x69 and initial 0xFF
```
Multilin comment should be witg /*

lordfolken commented on src/Device/Driver/LX_EOS/LXEosParser.cpp:
```diff
@@ -0,0 +1,186 @@
+
+    // Negative result can happen when the empty weight and/or crew weight
+    // setting in XCSoar is higher than in the vario
+    if (ballast_overload < 1)
```
This can be smaller than 1. Lxnav limits it to .8

lordfolken commented on NEWS.txt:
```diff
@@ -20,6 +20,7 @@ Version 7.42 - 2024/03/01
   - ignore unit for radio frequency in airspace files
 * devices
   - validate wind direction
+  - added driver for LX Navigation Eos57 variometer
```
The news entry should be on the unrelased version 7.43 not 7.42

lordfolken reviewed (dismissed):
I would be willing to merge this, as long as you support this driver. 
Would you be ready for merge?

lordfolken commented:
> I noticed one more issue:
> 
> If the Eos is connected together with another device (Flarm, for example), the altitude provided by the other dece is prioritized regardless of the order in the device list.
> 
> In other words: the info.ProvideBaroAltitudeTrue(altitude_true) is "weaker" than info.ProvidePressureAltitude(value).
> 
> Can this be considered a bug in XCSoar, or is it a feature?

A feature that doesn't consider your current usecase:
https://github.com/XCSoar/XCSoar/blob/57ab4a6934ebeeecb66740736a104c6fa115cb71/src/Computer/BasicComputer.cpp#L67-L85

jirimaier commented:
> > I noticed one more issue:
> > If the Eos is connected together with another device (Flarm, for example), the altitude provided by the other dece is prioritized regardless of the order in the device list.
> > In other words: the info.ProvideBaroAltitudeTrue(altitude_true) is "weaker" than info.ProvidePressureAltitude(value).
> > Can this be considered a bug in XCSoar, or is it a feature?
> 
> A feature that doesn't consider your current usecase:
> 
> https://github.com/XCSoar/XCSoar/blob/57ab4a6934ebeeecb66740736a104c6fa115cb71/src/Computer/BasicComputer.cpp#L67-L85

This shouldn't matter in the current implementation, because now it uses the QNH offset from LXWP3 to calculate standard pressure altitude and provides it using ProvidePressureAltitude. ProvideBaroAltitudeTrue is only used in the fallback case used for ERA and older EOS versions (where we don't know if the bug is present or not).

jirimaier commented:
> I would be willing to merge this, as long as you support this driver. Would you be ready for merge?

Yes, I am ready.

lordfolken commented:
* I renamed the driver to LxEos simply because it supports the 57 and 80 mm versions of the device?
* The Era I tested as well as i could... 
* In device Register.cpp you have to add the driver to the END of the enum, otherwise you mess up people's configurations
* rebased the code on current master for merge. 

@jirimaier  Do you agree?

lordfolken commented:
@jirimaier could you have a look at the past two comments?

lordfolken commented:
Woohoo! Way to go! :-) Thanks for sticking with it.
